enhancements/big-bundle-unpacking.md

```diff
@@ -0,0 +1,138 @@
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+To ensure Operator Bundles of arbitrary size can be unpacked and used in a Kubernetes cluster by OLM, bundle content will be partitioned across several ConfigMaps. This unpacking behavior will be abstracted behind a new Kubernetes API to allow for future extensibility.
+
+## Motivation
+
+Currently, OLM stages Operator Bundles within a cluster before installation by writing their content into ConfigMaps. There is a 1:1 relationship between bundle and ConfigMap; i.e. each bundle is unpacked to its own ConfigMap. A limitation of this staging method is that the maximum size of a ConfigMap for a given k8s cluster becomes the upper bound for the size of a bundle installed to it.
+
+### Goals
+
+- Enable OLM to install operators from bundles containing more content than fits in a single ConfigMap
+
+### Non-Goals
+
+- Increase the upper storage bound for API resources in general
+- Support bundles containing individual resources that surpass the upper storage bound of a cluster
+
+## Proposal
+
+A new Kubernetes resource kind will be introduced into the `porcelain.operators.coreos.com/v1` group version: `Bundle`. A `Bundle` resource will abstract the pulling and unpacking of an operator bundle from clients (OLM and cluster admins) and will feature the following fields:
+
+| ----- | ----------- |
+| spec.from | A union type which locates the bundle (e.g. a docker image digest) |
+| status.conditions | [Standard k8s conditions](https://github.com/kubernetes/apimachinery/blob/94222d04a59075a01fddedd696037db9e61db6e9/pkg/apis/meta/v1/types.go#L1367), surfacing info about the unpack job |
+| status.mediaType | Discriminates the type of bundle pulled (e.g. `registry+v1`) |
+
+```yaml
+apiVersion: porcelain.operators.coreos.com/v1alpha1
+kind: Bundle
+metadata:
+  name: my-operator
+  namespace: olm
+spec:
+  from:
+    type: Image
+    ref: quay.io/njhale/my-operator@sha256:123def...
+status:
+  mediaType: registry+v1
+  conditions:
+    type: UnpackFailed
+    status: True
+    reason: InvalidBundle
+    message: "no manifests detected in bundle"
+```
+
+Most importantly, a `Bundle` will expose a `/content` subresource which -- upon a successful unpackaging -- will return a [JSON encoded representation]() of that bundle.
+
+OLM will come equipped with a default aggregated API-server to provide the `porcelain` API group. The controller embedded in the default API-server will reconcile `Bundle` resources and spawn "unpack" `Job` resources similar to those currently employed, with a few key differences:
+
+- Unpack `Jobs` will now discretize bundle contents into chunks that are then stored in distinct `ConfigMaps`. With a sufficiently small chunk size, a bundle that is too large to fit into a single `ConfigMap` can be stored using several
```

Comment:
Is there an upper bound to the number of configmaps we are willing to create to shard the bundle contents across?

I'm asking because previously, if the bundle was too big for 1 configmap, OLM would fail. Now we are introducing the possibility of putting OLM in a bad state by packaging huge arbitrary manifests in a bundle and having dozens of configmaps store those manifests. Since this configmap data is stored on the cluster etcd, it could eventually start impacting the performance of the overall cluster.

Reply:
Good point. Let's break down the space complexity.

If we have _b_  bundles (all versions of all operators),  _c_ as the constant overhead per configmap (e.g. type, metadata, etc, content header), and _o_ as the number of installed bundes (where _o < b_ assuming singletons for simplicity), after unpacking _all_ bundles we should come to _c(b + o)_ -> _O(b)_ (keeping only the largest term).

I don't think this is any worse asymptotically than what we have today.

Reply:
That makes sense, thanks for the explanation. Maybe worth referring to in the enhancement

Reply:
I think the main point, which manages to elude the above complexity analysis, is that a single "bad" bundle may now claim an arbitrary, unbounded storage of a cluster's etcd. This needs to be considered from an access control perspective, since anyone with the capability of introducing a new bundle to a subscribed channel, can effectively choke an entire cluster.

It might be worthwhile to introduce a configurable knob (e.g., catalog-operator flag) with a reasonable default, that allows to cap the maximum size that a bundle may unpack onto a cluster.